src/statsmodels/formula.jl

```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
 end
 dropUnusedLevels!(x) = x
 
-function ModelFrame(trms::Terms, d::AbstractDataFrame)
+## whether or not a column of a particular type can be "promoted" to full rank
+## (only true of factors)
+can_promote(::PooledDataArray) = true
+can_promote(::Any) = false
+
+## Check for non-redundancy of columns.  For instance, if x is a factor with two
+## levels, it should be expanded into two columns in y~0+x but only one column
+## in y~1+x.  The default is the rank-reduced form (contrasts for n levels only
+## produce n-1 columns).  In general, an evaluation term x within a term
+## x&y&... needs to be "promoted" to full rank if y&... hasn't already been
+## included (either explicitly in the Terms or implicitly by promoting another
+## term like z in z&y&...).
+##
+## This function returns a boolean matrix that says whether each evaluation term
```

Comment:
"evaluation term" isn't the best expression to distinguish from "term". Maybe "model matrix column"? I think this could also apply to the `termnames` field of the contrasts type (-> "colnames"?).